plugin/pulp/plugin/download.py

```diff
@@ -0,0 +1,166 @@
+"""
+This module contains set of download APIs.
+"""
+import os
+import hashlib
+
+import requests
+from requests_futures.sessions import FuturesSession
+from requests.packages.urllib3.util import retry
+
+# TODO make config settings
+DEFAULT_CONNECT_TIMEOUT = 6.1
+DEFAULT_READ_TIMEOUT = 30
+DEFAULT_TRIES = 5
+DEFAULT_CONCURRENCY = 5
+# TODO Picked a bit out of a hat - I expect it should try to line up with the FS.
+DEFAULT_CHUNKSIZE = 1024 * 1024
+
+
+def default_callback_generator(filename=None, size=None, checksums=None):
+    """
+    Creates a closure to write a response to disk, checksum it, and verify
+    its size.
+
+    :param filename: The absolute path the response should be saved to.
+    :param size: The size, in bytes, that the file is expected to be; check is
+                 skipped if this value is not provided.
+    :param checksums: A dictionary of checksums to validate the file with; the
+                      key should be the hash algorithm and the value the
+                      expected hex digest.
+    """
+    # TODO potentially add a way to have this generate the entire checksum set.
+    # Maybe if there's {'algorithm': None}?
+    checksums = checksums or {}
+
+    def callback(self, response):
+        """
+        A callback for the requests_futures library.
+
+        :param resp: The requests response object
+        """
+        hashes = {algorithm: hashlib.new(algorithm) for algorithm in checksums}
+        with open(filename, mode='wb') as fp:
+            chunks = response.iter_content(chunk_size=DEFAULT_CHUNKSIZE)
+            for chunk in chunks:
+                for hash in hashes.values():
+                    hash.update(chunk)
+                fp.write(chunk)
+
+        # Validate that what we downloaded is what the user expects
+        # TODO add more reasonable exceptions
+        if size and size != os.stat(filename).st_size:
+            raise ValueError()
+
+        if checksums != {algorithm: hash.hexdigest() for algorithm, hash in hashes.items()}:
+            raise ValueError()
+
+    return callback
+
+
+class FileFuturesSession(FuturesSession):
+    """
+    This class implements the requests-futures API, with one  small adjustment. It provides
+    a default callback that writes a file to disk and validates its content. Once it
+    successfully returns, the file is ready to be imported into Pulp.
+
+    requests-futures allows you to queue up a large number of requests which are run
+    asynchonously on a pool of worker threads or processes. For examle:
+
+        >>> session = FileFuturesSession(max_workers=10)  # Create a pool of 10 threads
+        >>> requests = [
+        ...     {url='http://example.com/file1'},
```

Comment:
Note that as the implementation currently stands, this won't work. You'd need to do something like `((url,), {})` and then 

```
for args, kwargs in requests:
    session.get(*args, **kwargs)
```